capnp/lib/capnp.pyx

```diff
@@ -1936,194 +2010,145 @@ cdef class _CallContext:
+        return _promise_then(self, func, error_func, 1)
 
-        ret = _Promise()._init(self.thisptr.attach(capnp.makePyRefCounter(<PyObject *>args)), self)
-        self.is_consumed = True
-
-        return ret
-
-    cpdef cancel(self, numParents=1) except +reraise_kj_exception:
-        if numParents > 0 and hasattr(self._parent, 'cancel'):
-            self._parent.cancel(numParents - 1)
-
-        self.is_consumed = True
-        del self.thisptr
-        self.thisptr = NULL
+    cpdef cancel(self) except +reraise_kj_exception:
+        self.thisptr = Own[PyPromise]()
 
 
 cdef class _VoidPromise:
-    cdef VoidPromise * thisptr
-    cdef public bint is_consumed
-    cdef public object _parent
-    cdef _EventLoop _event_loop
+    cdef Own[VoidPromise] thisptr
 
-    def __init__(self):
-        self.is_consumed = True
-        self._event_loop = C_DEFAULT_EVENT_LOOP_GETTER()
 
-    cdef _init(self, VoidPromise other, parent=None):
-        self.is_consumed = False
-        self.thisptr = new VoidPromise(moveVoidPromise(other))
-        self._parent = parent
+    cdef _init(self, VoidPromise other):
+        self.thisptr = capnp.heap[VoidPromise](moveVoidPromise(other))
         return self
 
-    def __dealloc__(self):
-        del self.thisptr
-
     cpdef wait(self) except +reraise_kj_exception:
-        if self.is_consumed:
-            raise KjException(
-                "Promise was already used in a consuming operation. You can no longer use this Promise object")
-
-        helpers.waitVoidPromise(self.thisptr, deref(self._event_loop.thisptr).waitScope)
-
-        self.is_consumed = True
+        _promise_check_consumed(self)
+        cdef Own[VoidPromise] prom = move(self.thisptr) # Explicit move to not leave thisptr dangling
+        cdef _EventLoop loop = C_DEFAULT_EVENT_LOOP_GETTER()
+        with nogil:
+            prom.get().wait(deref(loop.waitScope))
 
-    cpdef then(self, func, error_func=None) except +reraise_kj_exception:
-        if self.is_consumed:
-            raise KjException(
-                "Promise was already used in a consuming operation. You can no longer use this Promise object")
+    async def a_wait(self):
+        """
+        Asyncio version of wait().
+        Required when using asyncio for socket communication.
 
-        argspec = None
-        try:
-            argspec = _inspect.getfullargspec(func)
-        except (TypeError, ValueError):
-            pass
-        if argspec:
-            args_length = len(argspec.args) if argspec.args else 0
-            defaults_length = len(argspec.defaults) if argspec.defaults else 0
-            if args_length - defaults_length != 0:
-                raise KjException('Function passed to `then` call must take no arguments')
+        Will still work with non-asyncio socket communication, but requires async handling of the function call.
+        """
+        # TODO: Is keeping a separate _VoidPromise class really worth it? Does it make things faster?
```

Comment:
I unfortunately, don't know the answer to this question.

Reply:
Yeah, I'm keeping it like it is for now. But in the future, someone might look into refactoring/simplifying the class hierarchy of Promises more.